File: FZU/PlanA/Phase 5/Practice/Codes/TEMPLATES/DP/DP.py
class DP:
    dp = None
    base_cases = None
    depend_func = None
    transit_func = None
    base_check = None

    def __init__(self, transit_func, depend_func, base_cases=None):
        self.transit_func = transit_func
        self.depend_func = depend_func
        self.dp = {}
        self.base_cases = set()
        if base_cases is not None:
            for base_case_state in base_cases.keys():
                self.set_base_case(state=base_case_state, dp_state=base_cases[base_case_state])

    def query(self, state, no_recur=False, from_cal=False):
        if self.base_check is not None:
            res = self.base_check(state)
            if res is not None:
                self.dp[state] = res
                return res
        if state in self.dp.keys() and self.dp[state] is not None:
            return self.dp[state]
        else:
            if no_recur:
                assert (not from_cal)
            return self.calculate(state, no_recur=no_recur)

    def set_base_check(self, base_check=None):
        self.base_check = base_check

    def set_base_case(self, state, dp_state):
        self.base_cases.add(state)
        self.dp[state] = dp_state

    def calculate(self, state, no_recur=False):
        depend_indices: dict = self.depend_func(state=state)
        depend_dp_dict = {}
        for depend_key in depend_indices.keys():
            depend_dp = self.query(
                state=depend_indices[depend_key],
                no_recur=no_recur,
                from_cal=True
            )
            depend_dp_dict[depend_key] = depend_dp
        result = self.transit_func(depend_dp_dict=depend_dp_dict, state=state)
        self.dp[state] = result
        return result

File: 复试/一志愿/Phase 5/Practice/Codes/TEMPLATES/DP/DP.cpp
#include <map>
#include <string>
#include <set>

using namespace std;

class DP
{
public:
    map<string, string> dp;
    set<string> base_cases;
    map<string, string> (*depend_func)(string);
    string (*transit_func)(map<string, string>, string);
    bool (*base_check)(string, string &);

    DP(
        map<string, string> (*depend_func)(string),
        string (*transit_func)(map<string, string>, string),
        map<string, string> base_cases = {})
    {
        this->transit_func = transit_func;
        this->depend_func = depend_func;
        this->dp = map<string, string>{};
        this->base_cases = set<string>{};
        this->base_check = nullptr;
        if (!base_cases.empty())
        {
            for (const auto &pair : base_cases)
            {
                string state = pair.first;
                string dp_state = pair.second;
                this->set_base_case(state, dp_state);
            }
        }
    }

    void set_base_case(string state, string dp_state)
    {
        this->base_cases.insert(state);
        this->dp[state] = dp_state;
    }

    void set_base_check(bool (*base_check)(string, string &) = nullptr)
    {
        this->base_check = base_check;
    }

    string query(string state)
    {
        string res;
        bool base = false;
        if (this->base_check != nullptr)
        {
            base = this->base_check(state, res);
        }
        if (this->dp.count(state) > 0)
        {
            base = true;
            res = this->dp[state];
        }
        if (!base)
        {
            res = this->calculate(state);
        }
        return res;
    }

    string calculate(string state)
    {
        string res;
        map<string, string> depend_indices = this->depend_func(state);
        map<string, string> depend_dp_dict;
        for (const auto &pair : depend_indices)
        {
            string depend_title = pair.first;
            string depend_state = pair.second;
            string depend_dp = this->query(depend_state);
            depend_dp_dict[depend_title] = depend_dp;
        }
        res = this->transit_func(depend_dp_dict, state);
        this->dp[state] = res;
        return res;
    }
};